File: frontend/src/services/authService.js
// Mạnh Hùng - HE200743
// Service xử lý các thao tác xác thực: đăng ký, xác minh email, đăng nhập bệnh nhân/nhân viên (2FA OTP),
// quên/đặt lại mật khẩu và đổi mật khẩu qua OTP.
// Tất cả các hàm đều gọi API backend thông qua axiosClient đã cấu hình sẵn base URL và token.
import axiosClient from '../api/axiosClient'

const authService = {
  // Đăng ký tài khoản bệnh nhân mới (chưa cấp token — cần xác minh email trước khi đăng nhập)
  register: (data) => axiosClient.post('/v1/auth/register', data),

  // Xác minh email bằng token nhận được qua liên kết trong email
  verifyEmail: (token) => axiosClient.get('/v1/auth/verify-email', { params: { token } }),

  // Gửi lại email xác minh cho tài khoản chưa kích hoạt
  resendVerification: (email) => axiosClient.post('/v1/auth/resend-verification', { email }),

  // Đăng nhập bệnh nhân bằng email-hoặc-số điện thoại và mật khẩu, trả về token JWT cùng thông tin người dùng
  login: (credentials) => axiosClient.post('/v1/auth/login', credentials),

  // Đăng nhập bằng tài khoản Google: gửi ID token lên backend để xác minh và tạo/đăng nhập tài khoản
  loginWithGoogle: (idToken) => axiosClient.post('/v1/auth/google', { idToken }),

  // Bước 1 đăng nhập nhân viên: xác thực email/mật khẩu, gửi mã OTP qua email (chưa cấp token)
  staffLogin: (credentials) => axiosClient.post('/v1/auth/staff/login', credentials),

  // Bước 2 đăng nhập nhân viên: xác minh mã OTP, nhận token JWT nếu hợp lệ
  staffVerifyOtp: (email, otp) => axiosClient.post('/v1/auth/staff/verify-otp', { email, otp }),

  // Yêu cầu quên mật khẩu: gửi email chứa liên kết đặt lại mật khẩu (nếu email tồn tại)
  forgotPassword: (email) => axiosClient.post('/v1/auth/forgot-password', { email }),

  // Đặt lại mật khẩu bằng token nhận được qua email
  resetPassword: (data) => axiosClient.post('/v1/auth/reset-password', data),

  // Bước 1 đổi mật khẩu cho người dùng đang đăng nhập (yêu cầu mật khẩu hiện tại và mật khẩu mới)
  changePassword: (data) => axiosClient.put('/v1/auth/change-password', data),

  // Bước 2 đổi mật khẩu: xác minh mã OTP xác nhận rồi áp dụng mật khẩu mới
  verifyChangePasswordOtp: (otp) => axiosClient.post('/v1/auth/change-password/verify-otp', { otp }),
}

export default authService

File: backend/src/main/java/com/ecms/service/AuthService.java
// Mạnh Hùng - HE200743
// Interface định nghĩa các hành vi xác thực người dùng: đăng ký, xác minh email, đăng nhập (bệnh nhân/nhân viên),
// xác thực OTP, quên/đặt lại mật khẩu, đổi mật khẩu và mở khóa tài khoản (admin).
package com.ecms.service;

import com.ecms.dto.request.*;
import com.ecms.dto.response.AuthResponse;

public interface AuthService {

    // Đăng ký tài khoản bệnh nhân mới ở trạng thái PENDING_VERIFICATION và gửi email xác minh
    void register(RegisterRequest request);

    // Xác minh email bằng token nhận qua link, kích hoạt tài khoản (chuyển sang ACTIVE)
    void verifyEmail(String token);

    // Gửi lại email xác minh cho tài khoản chưa được kích hoạt
    void resendVerification(ResendVerificationRequest request);

    // Đăng nhập bệnh nhân bằng email-hoặc-số điện thoại + mật khẩu, yêu cầu tài khoản đã ACTIVE
    AuthResponse login(LoginRequest request);

    // Đăng nhập/đăng ký bằng tài khoản Google: xác minh ID token, tìm hoặc tạo tài khoản với vai trò PATIENT
    AuthResponse loginWithGoogle(GoogleLoginRequest request);

    // Bước 1 đăng nhập nhân viên: xác thực email/mật khẩu rồi gửi mã OTP qua email, chưa cấp JWT
    void staffLogin(StaffLoginRequest request);

    // Bước 2 đăng nhập nhân viên: xác minh mã OTP, cấp JWT nếu hợp lệ
    AuthResponse staffVerifyOtp(StaffVerifyOtpRequest request);

    // Yêu cầu quên mật khẩu: luôn phản hồi giống nhau để tránh dò email tồn tại trong hệ thống
    void forgotPassword(ForgotPasswordRequest request);

    // Đặt lại mật khẩu bằng token nhận qua email
    void resetPassword(ResetPasswordRequest request);

    // Bước 1 đổi mật khẩu (đã đăng nhập): xác minh mật khẩu hiện tại rồi gửi mã OTP xác nhận
    void changePassword(String email, ChangePasswordRequest request);

    // Bước 2 đổi mật khẩu: xác minh mã OTP rồi áp dụng mật khẩu mới đã chờ xác nhận
    void verifyChangePasswordOtp(String email, VerifyChangePasswordOtpRequest request);

    // Mở khóa tài khoản nhân viên bị khóa do nhập sai mật khẩu quá số lần cho phép (chỉ ADMIN)
    void adminUnlockUser(Long userId);
}
